test/TestPoolCreation.ts

```diff
@@ -0,0 +1,221 @@
+
+      await expect(happyTokenPoolDeployed.fill_pool.apply(null, Object.values(invalidParams))).to.be.revertedWith(
+        getRevertMsg("Limit needs to be less than or equal to the total supply"),
+      );
+    });
+
+    it("Should throw error when the size of exchange_ratios does not correspond to exchange_addrs", async () => {
+      const invalidParams = {
+        ...creationParams,
+        exchange_ratios: creationParams.exchange_ratios.concat([4000, 1]),
+      };
+
+      await testTokenADeployed.approve(happyTokenPoolDeployed.address, creationParams.total_tokens);
+
+      await expect(happyTokenPoolDeployed.fill_pool.apply(null, Object.values(invalidParams))).to.be.revertedWith(
+        getRevertMsg("Size of ratios = 2 * size of exchange_addrs"),
+      );
+    });
+
+    it("Should throw error when tokens approved to spend is less than total_tokens", async () => {
+      const tokens_approved = ethers.utils.parseEther("1000");
+      const invalidParams = {
+        ...creationParams,
+        total_tokens: ethers.utils.parseEther("1001"),
+      };
+
+      await testTokenADeployed.approve(happyTokenPoolDeployed.address, tokens_approved);
+
+      await expect(happyTokenPoolDeployed.fill_pool.apply(null, Object.values(invalidParams))).to.be.reverted;
+    });
+
+    // TODO: check
+    // it("Should throw error when time is larger than 28 bits", async () => {
+    //   const invalidParams = {
+    //     ...creationParams,
+    //     start_time: 2 ** 28 - 1,
+    //     end_time: 2 ** 28 + 99,
+    //     lock_time: 0,
+    //   };
+
+    //   await testTokenADeployed.approve(happyTokenPoolDeployed.address, creationParams.total_tokens);
+
+    //   await expect(happyTokenPoolDeployed.fill_pool.apply(null, Object.values(invalidParams))).to.be.revertedWith(
+    //     getRevertMsg("Time is larger than 28 bits"),
+    //   );
+    // });
+
+    it("Should emit fillSuccess event correctly when a happyTokenPoolDeployed is filled", async () => {
+      const creatorAddress = await creator.getAddress();
+      const user_address = await ito_user.getAddress();
+
+      expect(creationParams.token_address).that.to.be.eq(testTokenADeployed.address);
+
+      const creatorBalanceBefore = await testTokenADeployed.balanceOf(creatorAddress);
+      const contractBalanceBefore = await testTokenADeployed.balanceOf(happyTokenPoolDeployed.address);
+
+      await testTokenADeployed.approve(happyTokenPoolDeployed.address, creationParams.total_tokens);
+      await happyTokenPoolDeployed.fill_pool.apply(null, Object.values(creationParams));
+      {
+        // filter with signature, should work
+        const logs = await ethers.provider.getLogs(happyTokenPoolDeployed.filters.FillSuccess());
+        const parsedLog = itoInterface.parseLog(logs[0]);
+        const result = parsedLog.args;
+        expect(result.total.toString()).that.to.be.eq(creationParams.total_tokens);
```

Comment:
Ethers.js provide a more convenient method for retrieving events: queryFilter(). See detail at: <https://docs.ethers.io/v5/api/contract/contract/#Contract-queryFilter>.  Moreover, typechain also helps us define th event type. You may try this method intergated with typechain, which could help simplify our code.